provider/core/src/datagen/data_conversion.rs

```diff
@@ -0,0 +1,64 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::prelude::*;
+use alloc::boxed::Box;
+use core::fmt;
+
+/// A trait that allows for converting between data payloads of different types.
+///
+/// These payloads will typically be some kind of erased payload, either with
+/// AnyMarker, BufferMarker, or SerializeMarker, where converting requires reifying the type.
+/// A type implementing [`DataConverter`] will essentially have a "registry" mapping keys to
+/// concrete marker types M, and reifying the input to a `DataPayload<M>`, performing some conversion
+/// or computation, and erasing the result to `DataPayload<MTo>`.
+///
+/// It will typically be implemented on data providers used in datagen.
+///
+/// The `impl_dyn_provider` is able to automatically implement this trait.
+pub trait DataConverter<MFrom: DataMarker, MTo: DataMarker> {
+    fn convert(
+        &self,
+        key: crate::ResourceKey,
+        from: DataPayload<MFrom>,
+    ) -> Result<DataPayload<MTo>, ReturnedPayloadError<MFrom>>;
+}
+
+impl<MFrom, MTo, P> DataConverter<MFrom, MTo> for Box<P>
+where
+    MFrom: DataMarker,
+    MTo: DataMarker,
+    P: DataConverter<MFrom, MTo> + ?Sized,
+{
+    fn convert(
+        &self,
+        key: crate::ResourceKey,
+        from: DataPayload<MFrom>,
+    ) -> Result<DataPayload<MTo>, ReturnedPayloadError<MFrom>> {
+        (**self).convert(key, from)
+    }
+}
+
+pub struct ReturnedPayloadError<M: DataMarker>(pub DataPayload<M>, pub DataError);
```

Comment:
It looks like you added this `ReturnedPayloadError` but didn't tie it in with the actual `downcast()` function.

I suggest moving this into the `any` module and then returning it from `downcast()`.

Reply:
Yes, the plan was to do it later, it's not just any that needs it.